libraries/plugins/voting_stat/voting_stat_plugin.cpp

```diff
@@ -0,0 +1,372 @@
+
+#include <graphene/voting_stat/voting_statistics_object.hpp>
+#include <graphene/voting_stat/voteable_statistics_object.hpp>
+#include <graphene/chain/worker_object.hpp>
+#include <graphene/chain/witness_object.hpp>
+#include <graphene/chain/committee_member_object.hpp>
+
+#include <graphene/protocol/config.hpp>
+
+#include <boost/signals2/connection.hpp>
+#include <boost/signals2/shared_connection_block.hpp>
+
+#include <memory>
+
+namespace graphene { namespace voting_stat {
+
+using namespace graphene::chain;
+using graphene::db::primary_index;
+
+namespace detail {
+
+class voting_stat_plugin_impl
+{
+public:
+   voting_stat_plugin_impl(voting_stat_plugin& _plugin) : _self( _plugin ){}
+   virtual ~voting_stat_plugin_impl(){}
+
+   boost::signals2::connection _on_voting_stake_calc_conn;
+   std::unique_ptr<boost::signals2::shared_connection_block> _on_voting_stake_calc_block;
+
+   uint16_t _maint_counter = 0;
+
+   /**
+    * plugin parameters
+    */
+   bool     _keep_objects_in_db     = true;
+   uint16_t _track_every_x_maint    = 12;
+   bool     _track_worker_votes     = true;
+   bool     _track_witness_votes    = true;
+   bool     _track_committee_votes  = true;
+
+   /**
+    * @brief callback for database::on_stake_calculated
+    *
+    * This function is triggered when the calculation of a stake for a given account is done inside the maintenance
+    * interval. It creates/updates the voting_statistics_object for a stake account. Optionally if a proxy is set
+    * in the stake account, also the voting_statistics_object for the proxy account is created/updated.
+    */
+   void on_stake_calculated(const account_object& stake_account, const account_object& proxy_account, uint64_t stake);
+
+   /**
+    * @brief callback for database::on_maintenance_begin
+    *
+    * Updates the block number to the one where the maintenance interval occurs and unblocks the
+    * database::on_stake_calculated signal, so that statistics objects can be created.
+    */
+   void on_maintenance_begin(uint32_t block_num);
+
+   /**
+    * @brief callback for database::on_maintenance_end
+    *
+    * Disconnects the on_stake_calculated callback and deletes all statistics objects if
+    * _voting_stat_delete_objects_after_interval=true
+    */
+   void on_maintenance_end();
+
+   void create_voteable_statistics_objects();
+   void delete_all_statistics_objects();
+
+   graphene::chain::database& database()
+   {
+      return _self.database();
+   }
+
+private:
+   voting_stat_plugin& _self;
+
+   uint32_t _maint_block;
+   bool     _create_voteable = false;
+};
+
+void voting_stat_plugin_impl::on_maintenance_begin(uint32_t block_num)
+{
+   if( !_keep_objects_in_db )
+      delete_all_statistics_objects();
+
+   if( _maint_counter == _track_every_x_maint )
+   {
+      _on_voting_stake_calc_block->unblock();
+      _maint_counter = 0;
+      _maint_block = block_num;
+      _create_voteable = true;
+   }
+   ++_maint_counter;
```

Comment:
This will not work properly. If there's a fork around the maintenance interval, the counter may be incremented multiple times for a single maintenance interval, and not consistently across all nodes.

Reply:
Hm now this is a bit a problem. But if I would just remove that (ignore the fact that we then can't control the number of objects we track) wouldn't there be the same problem?  That different objects will be pushed to es from different nodes? 

Is there some form of signal which could be applied `on_fork`? That would decrement the counter?

Reply:
No, but you can detect the switch to a fork in `applied_block` - block numbers should increase strictly monotonically, i. e. if the new block number is the same or lower than the previous you know that the chain is switching to a fork.

Reply:
Ah okay. Understood. Will rebuild that.

Reply:
I have still some problems with that. E.g. `on_maintenance_begin` is called and the plugin routine is performed. Now a fork appears, which leads to another maintenance interval.
AFAIU the plugin routine should now be applied to the fork maintenance.
So when something like that happens I need to revert the changes from the previous interval and apply the plugin routine to the fork interval. Please correct me if I am wrong.

Reply:
Actually knowing that the chain is switching to a fork is not enough, because in extreme cases a fork can span more than one maintenance interval.
A catch-all solution would be to store the counter in a database field instead of using an instance variable. Database entries are rolled back automatically when switching to a different fork.

Reply:
Okay I see. I'll try to figure that out. But in such cases there should also be problems with `es_objects` plugin I think. For example I have added a `forbid_delete_flag` which forbids the `es_objects` delete operation for `vote*_objects` (so it's possible to delete them in the plugin but still store them in es).  Meaning a fork appears, rerolls the objects, which should normally delete them (probably). But this doesn't happen. 

But in general I think I understood your solution. I will create a `db object` which keeps track of the counter.

Reply:
> Meaning a fork appears, rerolls the objects, which should normally delete them (probably). But this doesn't happen.

Correct. IIRC we decided that it doesn't matter because objects will eventually be overwritten by new objects with the same ID.

Reply:
Ah indeed, you are right. Okay then I think I got no further question for now. Thanks for the help.